### PR TITLE
feat: add `blas/base/dgemm`

### PR DESCRIPTION
**Pull request by Pranavchiku**
Progresses #2039.

## Description

> What is the purpose of this pull request?

This RFC proposes to add a routine to perform one of the matrix-matrix operation `C = α*op(A)*op(B) + β*C` where `op(A)` is one of the `op(A) = A`, or `op(A) = A^T`, `α` and `β` are scalars, `A`, `B`, and `C` are matrices, with op(A) an `M` by `K` matrix, op(B) a `K` by `N` matrix and C an `M` by `N` matrix as defined in [BLAS Level 3](http://www.netlib.org/blas/#_level_3) routines. Specifically adding `@stdlib/blas/base/dgemm` is proposed.

## Related Issues

> Does this pull request have any related issues?

This pull request:

-   progresses #2039. 

## Questions

> Any questions for reviewers of this pull request?

No.

## Other

> Any other information relevant to this pull request? This may include screenshots, references, and/or implementation notes.

No.

## Checklist

- [x] Add readme.md
- [x] Add package.json
- [x] Add lib
- [x] Add examples
- [x] Add ndarray implementation
- [x] Add docs
- [x] Add benchmark
- [x] Add test
- [x] Refactor implementation ( not sure if we can do this in here )

> Please ensure the following tasks are completed before submitting this pull request.

-   [x] Read, understood, and followed the [contributing guidelines][contributing].

* * *

@stdlib-js/reviewers

[contributing]: https://github.com/stdlib-js/stdlib/blob/develop/CONTRIBUTING.md

### CODE REVIEW

File: lib/node_modules/@stdlib/blas/base/dgemm/lib/base.js [hunk 1]
```diff
@@ -0,0 +1,259 @@
...
+				}
+				for ( l = 0; l < k; l++ ) {
+					tmp = alpha * B[ offsetB + ( l * LDB ) + j ];
+					for ( i = 0; i < m; i++ ) {
+						C[ offsetC + ( i * LDC ) + j ] += tmp * A[ offsetA + ( i * LDA ) + l ]; // eslint-disable-line max-len
+					}
+				}
+			}
+			return C;
+		}
```
Comment (Pranavchiku) [inline lib/node_modules/@stdlib/blas/base/dgemm/lib/base.js]: When you see this code, the only difference is accessing and writing matrices `A`, `B` and `C`. How about a utility function with API:

```javascript
function stdlib_blas_base_matrix_read( order, A, LDA, offsetA, colA, rowA )
```

Where `colA` and `rowA` are the column and row index of element to be fetched and based on this we can fetch value.

Similar API to write

```javascript
function stdlib_blas_base_matrix_write( order, A, LDA, offsetA, colA, rowA, value )
```

We can figure out variable and function naming if idea seems fine.

cc @kgryte

Comment (Pranavchiku) [reply lib/node_modules/@stdlib/blas/base/dgemm/lib/base.js]: I am not sure about cache optimisation in these cases. My basic intuition states that on generalising it for both orders we won't be able to optimise the `row-major` case ( reading and writing via cache ). I'll be happy to read any article / blog in these directions.

Comment (kgryte) [reply lib/node_modules/@stdlib/blas/base/dgemm/lib/base.js]: This won't work due to inefficient traversals and you'd have nested conditionals, thus making compiler optimizations more difficult. The primary way you address this is through loop interchange. See base ndarray utilities (e.g., `ndarray/base/assign`) for an example.

**Comment by Pranavchiku** [comment]
I pushed the changes I had in my branch, I'll let @aman-095 work on this and finish this PR. Thank you!

**Comment by kgryte** [comment]
@aman-095 Would you mind porting over the recent changes from https://github.com/stdlib-js/stdlib/pull/2742 to this PR? Thanks!

**Review by kgryte — APPROVED** [review]
LGTM. Thanks, @Pranavchiku and @aman-095.

**Pull request merged**